In [ ]:
import sys
import ctypes    # It is an external module which can help us make the data types from the c language

In [3]:
L = []
for i in range(50):
    print(i,sys.getsizeof(L))
    L.append(i)

0 56
1 88
2 88
3 88
4 88
5 120
6 120
7 120
8 120
9 184
10 184
11 184
12 184
13 184
14 184
15 184
16 184
17 248
18 248
19 248
20 248
21 248
22 248
23 248
24 248
25 312
26 312
27 312
28 312
29 312
30 312
31 312
32 312
33 376
34 376
35 376
36 376
37 376
38 376
39 376
40 376
41 472
42 472
43 472
44 472
45 472
46 472
47 472
48 472
49 472


From above we can see that first the value of the size was 56 then 88 after 4 size later the size becomes 120 then after its double it becomes 184

and this goes ---> So we can conclude that the list act as a dynamic array however instead of doubling everytime in python we add **8** extra spaces

In [ ]:


class MeraList: 
 def __init__(self):           # Creating a constructor
  self.size = 1                # size = Max amount of data which can be store
  self.n = 0                   # n = How much the space is filled currently

  # Creat a C type array with size = self.size

  self.A = self.__make__array(self.size)

 def __make__array(self,capacity):
  return (capacity*ctypes.py_object)()      # This code creates a Ctype array with size capacity ( static and referential array)


